File: download_gifs.py
#!/usr/bin/env python3
"""
Script para descargar GIFs desde MongoDB y guardarlos como archivos locales
Uso: python download_gifs.py
"""

import base64
import os
from pymongo import MongoClient

# Configuración
MONGO_URI = "mongodb+srv://angel_artu:Test123456@includesign.zz0yofi.mongodb.net/?retryWrites=true&w=majority&appName=inclusign"
DATABASE_NAME = "inclusign"

# Directorios de salida
ABECEDARIO_DIR = "inclusing_language_flutter/assets/gifs/abecedario"
GESTOS_DIR = "inclusing_language_flutter/assets/gifs/gestos"

def format_bytes(bytes):
    """Formatear bytes a formato legible"""
    if bytes < 1024:
        return f"{bytes} B"
    elif bytes < 1024 * 1024:
        return f"{bytes / 1024:.1f} KB"
    else:
        return f"{bytes / (1024 * 1024):.1f} MB"

def download_abecedario(collection, output_dir):
    """Descargar GIFs del abecedario"""
    print(f"\n[>>] Descargando GIFs del Abecedario...")

    documents = list(collection.find())
    print(f"   Encontrados {len(documents)} documentos")

    success_count = 0
    error_count = 0

    for doc in documents:
        try:
            letra = doc.get('letra') or doc.get('nombre')
            contenido_base64 = doc.get('contenido')

            if not letra or not contenido_base64:
                print(f"   [!] Documento sin letra o contenido")
                error_count += 1
                continue

            # Decodificar base64
            gif_bytes = base64.b64decode(contenido_base64)

            # Guardar archivo
            filename = f"{letra}.gif"
            filepath = os.path.join(output_dir, filename)

            with open(filepath, 'wb') as f:
                f.write(gif_bytes)

            print(f"   [OK] {filename} ({format_bytes(len(gif_bytes))})")
            success_count += 1

        except Exception as e:
            print(f"   [ERROR] {e}")
            error_count += 1

    print(f"   [RESUMEN] Exitosos: {success_count}, Errores: {error_count}")

def download_gestos(collection, output_dir):
    """Descargar GIFs de gestos"""
    print(f"\n[>>] Descargando GIFs de Gestos...")

    documents = list(collection.find())
    print(f"   Encontrados {len(documents)} documentos")

    success_count = 0
    error_count = 0

    for doc in documents:
        try:
            nombre = doc.get('nombre')
            contenido_base64 = doc.get('contenido')

            if not nombre or not contenido_base64:
                print(f"   [!] Documento sin nombre o contenido")
                error_count += 1
                continue

            # Normalizar nombre: "BUENOS DIAS" -> "BUENOS_DIAS"
            normalized_name = nombre.upper().replace(' ', '_')
            # Remover caracteres especiales
            normalized_name = ''.join(c for c in normalized_name if c.isalnum() or c == '_')

            # Decodificar base64
            gif_bytes = base64.b64decode(contenido_base64)

            # Guardar archivo
            filename = f"{normalized_name}.gif"
            filepath = os.path.join(output_dir, filename)

            with open(filepath, 'wb') as f:
                f.write(gif_bytes)

            print(f"   [OK] {filename} ({format_bytes(len(gif_bytes))})")
            success_count += 1

        except Exception as e:
            print(f"   [ERROR] Error en '{nombre}': {e}")
            error_count += 1

    print(f"   [RESUMEN] Exitosos: {success_count}, Errores: {error_count}")

def count_files(directory):
    """Contar archivos .gif en un directorio"""
    try:
        return len([f for f in os.listdir(directory) if f.endswith('.gif')])
    except:
        return 0

def main():
    print("[*] Conectando a MongoDB...")

    # Conectar a MongoDB
    client = MongoClient(MONGO_URI)
    db = client[DATABASE_NAME]

    print(f"[OK] Conectado a MongoDB")
    print(f"[*] Base de datos: {DATABASE_NAME}")

    # Crear directorios si no existen
    os.makedirs(ABECEDARIO_DIR, exist_ok=True)
    os.makedirs(GESTOS_DIR, exist_ok=True)

    try:
        # Descargar abecedario
        abecedario_collection = db['Abecedario']
        download_abecedario(abecedario_collection, ABECEDARIO_DIR)

        # Descargar gestos
        gestos_collection = db['Gestos']
        download_gestos(gestos_collection, GESTOS_DIR)

    except Exception as e:
        print(f"\n[ERROR] Error general: {e}")
    finally:
        client.close()
        print(f"\n[OK] Descarga completa!")
        print(f"[*] Resumen:")
        print(f"   - Abecedario: {count_files(ABECEDARIO_DIR)} archivos")
        print(f"   - Gestos: {count_files(GESTOS_DIR)} archivos")

if __name__ == "__main__":
    main()

File: download_gifs.dart
import 'dart:io';
import 'dart:convert';
import 'package:mongo_dart/mongo_dart.dart';

/// Script para descargar GIFs desde MongoDB y guardarlos como archivos locales
///
/// Uso: dart run download_gifs.dart
void main() async {
  // Configuración de MongoDB
  final mongoUri = 'mongodb+srv://angel_artu:Test123456@includesign.zz0yofi.mongodb.net/?retryWrites=true&w=majority&appName=inclusign';
  final databaseName = 'inclusign';

  print('🔄 Conectando a MongoDB...');

  final db = await Db.create(mongoUri);
  await db.open();

  print('✅ Conectado a MongoDB');
  print('📚 Base de datos: $databaseName');

  // Seleccionar base de datos
  final database = db.useDb(databaseName);

  // Crear directorios si no existen
  final abecedarioDir = Directory('inclusing_language_flutter/assets/gifs/abecedario');
  final gestosDir = Directory('inclusing_language_flutter/assets/gifs/gestos');

  if (!abecedarioDir.existsSync()) {
    abecedarioDir.createSync(recursive: true);
  }
  if (!gestosDir.existsSync()) {
    gestosDir.createSync(recursive: true);
  }

  print('\n📥 Descargando GIFs del Abecedario...');
  await downloadAbecedario(database, abecedarioDir);

  print('\n📥 Descargando GIFs de Gestos...');
  await downloadGestos(database, gestosDir);

  await db.close();
  print('\n✅ ¡Descarga completa!');
  print('📊 Resumen:');
  print('   - Abecedario: ${countFiles(abecedarioDir)} archivos');
  print('   - Gestos: ${countFiles(gestosDir)} archivos');
}

/// Descargar GIFs del abecedario
Future<void> downloadAbecedario(Db database, Directory outputDir) async {
  final collection = database.collection('Abecedario');
  final documents = await collection.find().toList();

  print('   Encontrados ${documents.length} documentos en Abecedario');

  int successCount = 0;
  int errorCount = 0;

  for (var doc in documents) {
    try {
      final letra = doc['letra'] ?? doc['nombre'];
      final contenidoBase64 = doc['contenido'];

      if (letra == null || contenidoBase64 == null) {
        print('   ⚠️  Documento sin letra o contenido: ${doc['_id']}');
        errorCount++;
        continue;
      }

      // Decodificar base64
      final bytes = base64Decode(contenidoBase64);

      // Guardar archivo
      final fileName = '$letra.gif';
      final file = File('${outputDir.path}/$fileName');
      await file.writeAsBytes(bytes);

      print('   ✅ $fileName (${_formatBytes(bytes.length)})');
      successCount++;
    } catch (e) {
      print('   ❌ Error procesando documento: $e');
      errorCount++;
    }
  }

  print('   📊 Exitosos: $successCount, Errores: $errorCount');
}

/// Descargar GIFs de gestos
Future<void> downloadGestos(Db database, Directory outputDir) async {
  final collection = database.collection('Gestos');
  final documents = await collection.find().toList();

  print('   Encontrados ${documents.length} documentos en Gestos');

  int successCount = 0;
  int errorCount = 0;

  for (var doc in documents) {
    try {
      final nombre = doc['nombre'];
      final contenidoBase64 = doc['contenido'];

      if (nombre == null || contenidoBase64 == null) {
        print('   ⚠️  Documento sin nombre o contenido: ${doc['_id']}');
        errorCount++;
        continue;
      }

      // Normalizar nombre: "BUENOS DIAS" -> "BUENOS_DIAS"
      final normalizedName = (nombre as String)
          .toUpperCase()
          .replaceAll(' ', '_')
          .replaceAll(RegExp(r'[^A-Z0-9_]'), '');

      // Decodificar base64
      final bytes = base64Decode(contenidoBase64);

      // Guardar archivo
      final fileName = '$normalizedName.gif';
      final file = File('${outputDir.path}/$fileName');
      await file.writeAsBytes(bytes);

      print('   ✅ $fileName (${_formatBytes(bytes.length)})');
      successCount++;
    } catch (e) {
      print('   ❌ Error procesando "$doc['nombre']": $e');
      errorCount++;
    }
  }

  print('   📊 Exitosos: $successCount, Errores: $errorCount');
}

/// Contar archivos .gif en un directorio
int countFiles(Directory dir) {
  return dir
      .listSync()
      .where((entity) => entity is File && entity.path.endsWith('.gif'))
      .length;
}

/// Formatear bytes a formato legible
String _formatBytes(int bytes) {
  if (bytes < 1024) return '$bytes B';
  if (bytes < 1024 * 1024) return '${(bytes / 1024).toStringAsFixed(1)} KB';
  return '${(bytes / (1024 * 1024)).toStringAsFixed(1)} MB';
}
